std/machines/hash/poseidon_bb.asm

```diff
@@ -0,0 +1,204 @@
+    pol constant C_15 = [385833350, 3792692, 1072385714, 1194261137, 226255882, 368055032, 163075143, 574924181, 1162051737, 589145361, 75983087, 89570374, 1565702224, 315099133, 1862294673, 715849472, 25972299, 1529251108, 698696249, 932100232, 445083324, 0]*;
+    let C = [C_0, C_1, C_2, C_3, C_4, C_5, C_6, C_7, C_8, C_9, C_10, C_11, C_12, C_13, C_14, C_15];
+
+    // State of the Poseidon permutation (8 rate elements and 4 capacity elements)
+    pol commit state[STATE_SIZE];
+
+    // The first OUTPUT_SIZE elements of the *final* state
+    // (constrained to be constant within the block and equal to parts of the state in the last row)
+    pol commit output[OUTPUT_SIZE];
+
+    // Add round constants
+    // TODO should these be intermediate?
+    let a = array::zip(state, C, |state, C| state + C);
+
+    // Compute S-Boxes (x^7) (using a degree bound of 3)
+    col witness x3[STATE_SIZE];
+    array::zip(x3, array::map(a, |a| a * a * a), |x3, expected| x3 = expected);
+    col witness x7[STATE_SIZE];
+    array::zip(x7, array::zip(x3, a, |x3, a| x3 * x3 * a), |x7, expected| x7 = expected);
+
+    // Apply S-Boxes on the first element and otherwise if it is a full round.
+    let b: expr[] = array::new(STATE_SIZE, |i| if i == 0 {
+        x7[i]
+    } else {
+        PARTIAL * (a[i] - x7[i]) + x7[i]
+    });
+
+    // The MDS matrix from Plonk3 implementation
+    let M = [
+        [1, 1, 51, 1, 11, 17, 2, 1, 101, 63, 15, 2, 67, 22, 13, 3],
+        [3, 1, 1, 51, 1, 11, 17, 2, 1, 101, 63, 15, 2, 67, 22, 13],
+        [13, 3, 1, 1, 51, 1, 11, 17, 2, 1, 101, 63, 15, 2, 67, 22],
+        [22, 13, 3, 1, 1, 51, 1, 11, 17, 2, 1, 101, 63, 15, 2, 67],
+        [67, 22, 13, 3, 1, 1, 51, 1, 11, 17, 2, 1, 101, 63, 15, 2],
+        [2, 67, 22, 13, 3, 1, 1, 51, 1, 11, 17, 2, 1, 101, 63, 15],
+        [15, 2, 67, 22, 13, 3, 1, 1, 51, 1, 11, 17, 2, 1, 101, 63],
+        [63, 15, 2, 67, 22, 13, 3, 1, 1, 51, 1, 11, 17, 2, 1, 101],
+        [101, 63, 15, 2, 67, 22, 13, 3, 1, 1, 51, 1, 11, 17, 2, 1],
+        [1, 101, 63, 15, 2, 67, 22, 13, 3, 1, 1, 51, 1, 11, 17, 2],
+        [2, 1, 101, 63, 15, 2, 67, 22, 13, 3, 1, 1, 51, 1, 11, 17],
+        [17, 2, 1, 101, 63, 15, 2, 67, 22, 13, 3, 1, 1, 51, 1, 11],
+        [11, 17, 2, 1, 101, 63, 15, 2, 67, 22, 13, 3, 1, 1, 51, 1],
+        [1, 11, 17, 2, 1, 101, 63, 15, 2, 67, 22, 13, 3, 1, 1, 51],
+        [51, 1, 11, 17, 2, 1, 101, 63, 15, 2, 67, 22, 13, 3, 1, 1],
+        [1, 51, 1, 11, 17, 2, 1, 101, 63, 15, 2, 67, 22, 13, 3, 1]
+    ];
+
+
+    // Multiply with MDS Matrix
+    let dot_product = |v1, v2| array::sum(array::zip(v1, v2, |v1_i, v2_i| v1_i * v2_i));
+    let c = array::map(M, |M_row_i| dot_product(M_row_i, b));
+
+    // Copy c to state in the next row
+    array::zip(state, c, |state, c| (state' - c) * (1-LAST) = 0);
+
+    // In the last row, the first OUTPUT_SIZE elements of the state should equal output
+    array::zip(output, state, |output, state| LASTBLOCK * (output - state) = 0);
```

Comment:
Please apply this fix accordingly, otherwise it will break once #1806 is merged:
https://github.com/powdr-labs/powdr/pull/1806/commits/12ffb9f6121dc034ea7004e6b1d658172844c331